scheduler/utilization_based_host_allocator.go

```diff
@@ -0,0 +1,248 @@
+			msg := "more than 1 distro sent to UtilizationBasedHostAllocator"
+			grip.Emergency(message.Fields{
+				"message": msg,
+				"distros": hostAllocatorData.distros,
+			})
+			return nil, errors.New(msg)
+		}
+
+		// actual calculation logic is here
+		newHosts, err := numNewHosts(ctx, d, hostAllocatorData.taskQueueItems[name],
+			hostAllocatorData.existingDistroHosts[name], hostAllocatorData.freeHostFraction)
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "error calculating hosts for distro %s", name)
+		}
+		newHostsNeeded[name] = newHosts
+		count++
+	}
+
+	grip.Info(message.Fields{
+		"runner":        RunnerName,
+		"num_new_hosts": newHostsNeeded[distroName],
+		"message":       "requsting new hosts",
+	})
+
+	return newHostsNeeded, nil
+}
+
+// Calculate the number of hosts needed by taking the total task scheduled task time
+// and dividing it by the target duration. Request however many hosts are needed to
+// achieve that minus the number of free hosts
+func numNewHosts(ctx context.Context, d distro.Distro, taskQueue []model.TaskQueueItem,
+	existingHosts []host.Host, freeHostFraction float64) (int, error) {
+
+	if !d.IsEphemeral() {
+		return 0, nil
+	}
+	numNewHosts := 0
+
+	// allocate 1 host per task that is longer than the max duration
+	newTaskQueue, hostsForLongTasks := calcHostsForLongTasks(taskQueue, MaxDurationPerDistroHost)
+
+	// determine the total expected running time of scheduled tasks
+	scheduledTasksDuration := calcScheduledTasksDuration(newTaskQueue)
+
+	// determine how many free hosts we have that are already up
+	numFreeHosts, err := calcExistingFreeHosts(existingHosts, freeHostFraction, MaxDurationPerDistroHost)
+	if err != nil {
+		return numNewHosts, nil
+	}
+
+	// calculate how many new hosts are needed (minus the hosts for long tasks)
+	numNewHosts = calcNewHostsNeeded(scheduledTasksDuration, MaxDurationPerDistroHost, numFreeHosts, hostsForLongTasks)
+
+	// don't start more hosts than new tasks. This can happen if the task queue is mostly long tasks
+	if numNewHosts > len(taskQueue) {
+		numNewHosts = len(taskQueue)
+	}
+	// enforce the max hosts cap
+	if numNewHosts+len(existingHosts) > d.PoolSize {
+		numNewHosts = d.PoolSize - len(existingHosts)
+	}
+
+	// don't return negatives - can get here if over max hosts
+	if numNewHosts < 0 {
+		numNewHosts = 0
+	}
+
+	// alert if the distro is underwater
+	if d.PoolSize < 1 {
+		return 0, errors.Errorf("unable to plan hosts for distro %s due to pool size of %d", d.Id, d.PoolSize)
+	}
+	underWaterAlert := message.Fields{
+		"provider":  d.Provider,
+		"distro":    d.Id,
+		"runtime":   scheduledTasksDuration,
+		"runner":    RunnerName,
+		"message":   "distro underwater",
+		"num_hosts": len(existingHosts),
+		"max_hosts": d.PoolSize,
+	}
+	avgMakespan := scheduledTasksDuration / time.Duration(d.PoolSize)
+	grip.AlertWhen(avgMakespan > dynamicDistroRuntimeAlertThreshold, underWaterAlert)
+
+	// log scheduler stats
+	grip.Info(message.Fields{
```

Comment:
should we also put this in the event log?

Reply:
line 98 of wrapper.go already logs in the event log. Did you want to do something in addition to that?